src/librustc_trans/adt.rs

```diff
@@ -162,458 +161,232 @@ pub fn represent_type<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
+        (I1, true) => tcx.types.i8,
+        (I8, true) => tcx.types.i8,
+        (I16, true) => tcx.types.i16,
+        (I32, true) => tcx.types.i32,
+        (I64, true) => tcx.types.i64,
+    }
+}
+
+fn convert_layout_struct<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                    from: &layout::Struct,
+                                    size: layout::Size,
+                                    variant: Option<ty::VariantDef<'tcx>>,
+                                    substs: &Substs<'tcx>) -> Struct<'tcx> {
+    let fields = if let Some(v) = variant {
+        v.fields.iter().map(|field| {
+            monomorphize::field_ty(tcx, substs, field)
+        }).collect::<Vec<_>>()
+    } else {
+        Vec::default()
+    };
+    Struct {
+        size: size.bytes(),
+        sized: from.sized,
+        align: from.align.abi() as u32,
+        packed: from.packed,
+        fields: fields
     }
 }
 
 fn get_cases<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
                        adt: ty::AdtDef<'tcx>,
                        substs: &Substs<'tcx>)
-                       -> Vec<Case<'tcx>> {
+                       -> Vec<Vec<Ty<'tcx>>> {
     adt.variants.iter().map(|vi| {
-        let field_tys = vi.fields.iter().map(|field| {
+        vi.fields.iter().map(|field| {
             monomorphize::field_ty(tcx, substs, field)
-        }).collect();
-        Case { discr: Disr::from(vi.disr_val), tys: field_tys }
+        }).collect()
     }).collect()
 }
 
-fn mk_struct<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
-                       tys: &[Ty<'tcx>], packed: bool,
-                       scapegoat: Ty<'tcx>)
-                       -> Struct<'tcx> {
-    let sized = tys.iter().all(|&ty| type_is_sized(cx.tcx(), ty));
-    let lltys : Vec<Type> = if sized {
-        tys.iter().map(|&ty| type_of::sizing_type_of(cx, ty)).collect()
-    } else {
-        tys.iter().filter(|&ty| type_is_sized(cx.tcx(), *ty))
-           .map(|&ty| type_of::sizing_type_of(cx, ty)).collect()
-    };
-
-    ensure_struct_fits_in_address_space(cx, &lltys[..], packed, scapegoat);
-
-    let llty_rec = Type::struct_(cx, &lltys[..], packed);
-    Struct {
-        size: machine::llsize_of_alloc(cx, llty_rec),
-        align: machine::llalign_of_min(cx, llty_rec),
-        sized: sized,
-        packed: packed,
-        fields: tys.to_vec(),
-    }
-}
-
-#[derive(Debug)]
-struct IntBounds {
-    slo: i64,
-    shi: i64,
-    ulo: u64,
-    uhi: u64
+//This is used to determine the fields of the unsized variant of an enum.
+fn compute_null_fields<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
+                                 adt: ty::AdtDef<'tcx>,
+                                 substs: &Substs<'tcx>,
+                                 nndiscr: u64) -> Vec<Ty<'tcx>> {
+    let cases = get_cases(cx.tcx(), adt, substs);
+    assert_eq!(cases.len(), 2); //More than two variants is not the null pointer optimization.
+    cases[1-(nndiscr  as usize)].clone()
```

Comment:
Eventually `get_cases` should be avoided, IMO, and this would look at the variant directly.